### PR TITLE
Edit constructor page

### LINKED ISSUE #5462: Reorder constructors and fields in code example on `Constructors` page
### Page URL

https://dart.dev/language/constructors.html

### Page source

https://github.com/dart-lang/site-www/tree/main/src/language/constructors.md

### Describe the problem

The examples on the ‘Constructors’ page (https://dart.dev/language/constructors) currently place fields before constructors. This is inconsistent with the Dart and Flutter style guides, which recommend placing constructors first in a class. This inconsistency could potentially lead to confusion for developers, especially those who frequently switch between Dart and Flutter projects.

### Expected fix

Reorder the constructors in all examples on the ‘Constructors’ page to align with the Dart and Flutter style guides. Specifically, adhere to the guidelines from ‘sort_constructors_first’ (https://dart.dev/tools/linter-rules/sort_constructors_first) and ‘Constructors come first in a class’ (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#constructors-come-first-in-a-class).

### Additional context

This issue was raised following a pull request (https://github.com/dart-lang/site-www/pull/5452) that aimed to align the examples with the official style guides. The discussion in the pull request highlighted the tradition of placing fields before constructors in Dart, which is more familiar to developers from other programming languages. However, given the close relationship between Dart and Flutter, it’s worth considering the impact of these stylistic differences on the developer experience. This issue aims to gather feedback from the community on this matter.

### I would like to fix this problem.

- [X] I will try and fix this problem on dart.dev.

### PR DESCRIPTION
**Pull request by atsansone**
Fixes #5462

### CODE REVIEW

**Comment by MaryaBelanger** [comment]
I think the conversation of whether this style needs to be enforced across all our documentation should be a conversation involving the wider team. This feels like more of a design decision than a docs decision. Yes, we have a lint `sort_constructors_first` but it's personal choice, not even part of the recommended set. And we don't have an effective dart entry as far as I can tell.

File: src/content/language/constructors.md [hunk 1]
```diff
@@ -34,17 +34,17 @@ class Point {
 The `this` keyword refers to the current instance.
```
[the remark this answers is not in the document]
Comment (atsansone) [reply src/content/language/constructors.md:34]: **praise:** Good catch!

File: src/content/language/constructors.md [hunk 2]
```diff
@@ -271,23 +256,20 @@ Point.withAssert(this.x, this.y) : [!assert(x >= 0)!] {
...
+  Point(this.x, this.y) : distanceFromOrigin = sqrt(x * x + y * y);
+
   final double x;
   final double y;
   final double distanceFromOrigin;
-
-  Point(double x, double y)
-      : x = x,
-        y = y,
-        distanceFromOrigin = sqrt(x * x + y * y);
```
[the remark this answers is not in the document]
Comment (atsansone) [reply src/content/language/constructors.md:290]: That's my guess. This copy precedes a 2016 site migration, so I have no idea.

File: examples/misc/lib/language_tour/classes/point.dart [hunk 3]
```diff
@@ -8,11 +8,6 @@ const double yOrigin = 0;
 
 // #docregion class-with-distance-to, constructor-initializer
 class Point {
-  final double x;
-  final double y;
-
-  // Sets the x and y instance variables
-  // before the constructor body runs.
```
[the remark this answers is not in the document]
Comment (atsansone) [reply examples/misc/lib/language_tour/classes/point.dart:15]: I re-added it, but it doesn't appear anywhere that I could see after.

**Comment by atsansone** [comment]
That's fine. Maybe see if @eernstg has an opinion?

**Comment by eernstg** [comment]
> an opinion?

Not really. ;-)

I don't think any particular ordering dominates the Dart software universe, and nothing is supported explicitly by the style guide or similar documents (that I know of). According to [directives_ordering](https://dart.dev/tools/linter-rules/directives_ordering), imports should be sorted by kind (platform, package, relative) and then alphabetically, following [Effective Dart](https://dart.dev/effective-dart/style#ordering). It probably makes sense to extrapolate this to class member declarations (and mixin member declarations, etc.) This means that we'd have all the constructors together, all static members together, etc. It's still a non-trivial decision whether each group is sorted alphabetically, or it is organized "semantically" in some sense, or even unsorted/random.

I don't think the community has agreed on much more: Perhaps constructors [should go first](https://dart.dev/tools/linter-rules/sort_constructors_first), perhaps instance variables should go first (this seems to be quite common), and who knows whether static members should come before instance members or vice versa, and who knows whether we'd put all the instance members into one group, or keep getters and setters in one group and methods in another group (same for static, by the way). It would also be possible (and it seems consistent to me) to treat instance variables exactly like the getters/setters that they introduce, but I don't think that's common at all.

@munificent, do you think there's anything like a canonical ordering of class member declarations, or even a particular style which is more common than others?

**Comment by atsansone** [comment]
FYI: I'm not invested in the ordering myself. I thought that the issue submitter had a point and made changes accordingly. If you feel that a definitive stance is premature, I'm happy to shelve this.

**Comment by natebosch** [comment]
I do not think we should make any specific recommendation. When this has been discussed internally (where we have an even higher impetus for consistency) we have decided against adapting a fixed ordering. There can be strong reasons to order members in different ways based on the ways the members related to eachother.

If a class is so big that rules like this feel necessary to avoid getting lost in the source code, then it should probably be split up.

The flutter repo does specify this, but the flutter repo style guide is not written with the intention of being a recommendation for a wider audience like the `package:lints` rule sets or Effective Dart.

**Comment by MaryaBelanger** [comment]
@atsansone Let's close this PR and #5452 based on [Nate's comment](https://github.com/dart-lang/site-www/pull/5718#issuecomment-2067134299) (and [Parker's older one](https://github.com/dart-lang/site-www/pull/5452#pullrequestreview-1808656101)). I don't think it's meaningful to standardize for Dart, and I strongly agree with Parker's point:

> This _[Dart's tendency to place field declarations first despite Flutter style]_ results in some clarity about the properties of a class but more importantly is likely more familiar to users of other related programming languages.

**Comment by atsansone** [comment]
Closing this issue as Won't fix.

**Comment by Turskyi** [comment]
Hello @atsansone, @MaryaBelanger, @eernstg, and @natebosch,

I've been closely following the discussions around the Dart documentation style and have noticed the closure of my original issue #5462 and pull request #5452. I appreciate the team's consideration of various viewpoints.

I'm encouraged by the move towards greater consistency in the documentation, as seen in PR #5718, which closely aligns with the changes I proposed in #5452. Open-source thrives on transparent and inclusive collaboration, and I would have valued the opportunity to be part of the conversation, especially to understand the team's perspective more deeply.

I'd like to highlight a point from my [comment](https://github.com/dart-lang/site-www/pull/5452#issuecomment-1884191443) about the placement of fields and constructors, supported by examples from Java and Dart. This practice is intuitive and aligns with styles in other languages, potentially benefiting developers.

Here are the examples for clarity:
In Java:

```Java
public class Book {
    private String title;

    public Book(String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }
}
```

And in Dart:

```Dart
class Book {
  Book(this.title);

  String title;
}
```
In these examples, the practice of placing constructors before properties, such as getters and setters, is common and aligns with styles seen in other programming languages. This approach can make the code more intuitive for developers. In the Dart class, ‘title’ serves not just as a field but also implicitly as a getter and setter, which is why it is positioned after the constructor, following the idiomatic use of Dart’s syntactic sugar for more concise code.

While I respect the decision to close the issue as “Won’t fix,” I believe that my refutation of the quoted point was not fully considered in the discussion. It’s important for the community to have a comprehensive understanding of different perspectives before reaching a conclusion.

I hope that we can revisit this topic with a broader scope, perhaps involving more community members, to ensure all viewpoints are represented, leading to a more informed decision-making process.

I’m open to adapting my contributions based on the team’s feedback and the community’s consensus. I look forward to continuing to contribute to the Dart community and would appreciate any feedback on this message.

Thank you for your time.

**Comment by eernstg** [comment]
> I hope that we can revisit this topic

I think it will be discussed in the future. You could take a look at https://github.com/dart-lang/linter/issues/2367, perhaps voting for it and/or contributing to the discussion. At this time I don't think anyone has created a set of rules which is obviously the way to go, and it doesn't make much sense to enforce a set of rules for anyone who might have good reasons to do something else (e.g., ordering instance variable declarations according to their mutual relationships and semantics). So the next step would surely be to build consensus around a particular approach.

**Pull request closed without merging**